### PR TITLE
Handle impl Trait more correctly

### PR DESCRIPTION
**Pull request by flodiebold**
When calling a function, argument-position impl Trait is transparent; same for return-position impl Trait when inside the function. 

So in these cases, we need to represent that type not by `Ty::Opaque`, but by a type variable that can be unified with whatever flows into there.

### CODE REVIEW

File: crates/ra_hir_ty/src/infer.rs [hunk 1]
```diff
@@ -274,6 +274,28 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
         self.normalize_associated_types_in(ty)
     }
 
+    /// Replaces `impl Trait` in `ty` by type variables and obligations for
+    /// those variables. This is done for function arguments when calling a
+    /// function, and for return types when inside the function body, i.e. in
+    /// the cases where the `impl Trait` is 'transparent'. In other cases, `impl
+    /// Trait` is represented by `Ty::Opaque`.
+    fn insert_vars_for_impl_trait(&mut self, ty: Ty) -> Ty {
```
Comment (matklad) [inline crates/ra_hir_ty/src/infer.rs]: Maybe mark?

Comment (flodiebold) [reply crates/ra_hir_ty/src/infer.rs]: Done

File: crates/ra_hir_ty/src/tests/traits.rs [hunk 2]
```diff
@@ -1,7 +1,10 @@
-use super::{infer, type_at, type_at_pos};
-use crate::test_db::TestDB;
 use insta::assert_snapshot;
+
 use ra_db::fixture::WithFixture;
+use test_utils::covers;
+
+use super::{infer, infer_with_mismatches, type_at, type_at_pos};
+use crate::test_db::TestDB;
```
Comment (ice1000) [inline crates/ra_hir_ty/src/tests/traits.rs]: Wow! Are you using intellij-rust to develop this? 😉

Comment (flodiebold) [reply crates/ra_hir_ty/src/tests/traits.rs]: You mean because of the reordering? I just did it by hand, I use Emacs ;)

Comment (ice1000) [reply crates/ra_hir_ty/src/tests/traits.rs]: Yep, and it's surprising 😅 you mean Emacs with rust-analyzer right?

Comment (flodiebold) [reply crates/ra_hir_ty/src/tests/traits.rs]: Yes

**Pull request merged**